### ISSUE TITLE
RxJava Observables with side effects causing concurrency hazards

### ISSUE THREAD

**Issue by davidjirovec** [bitrich-info/xchange-stream#402]
Related to https://github.com/bitrich-info/xchange-stream/pull/387

https://github.com/bitrich-info/xchange-stream/blob/8ac7f030a949dd245e8c32c79622eb10b06757d5/xchange-binance/src/main/java/info/bitrich/xchangestream/binance/BinanceStreamingMarketDataService.java#L268

Orderbook which is modified here can be iterated downstream in the same time which causes ConcurrentModificationException. I believe clean solution to this is not to add synchronized as in https://github.com/bitrich-info/xchange-stream/pull/387 but to use Observables without side effects and pass immutable objects, or at least when modification of mutable object is needed, create a copy.

**Comment by badgerwithagun** [comment]
Yes, this problem is endemic. I've never tried because there's so much of this sort of code that to remove the downstream protections in my app would require me to find and eliminate every single one of these smells, and there are too many.
